### ISSUE TITLE
[BUG] Specviz: Fix app.py so subset can apply to multiple data

### ISSUE THREAD

**Issue by javerbukh** [spacetelescope/jdaviz#1921]
**Describe the bug**
TLDR: Core methods in app.py assume only one data is loaded at a time, or a subset is applied to only one data at a time.

There are a few issues that this ticket will be pointing out. I will try to link to the code and provide examples where I can

`app.get_data_from_viewer()` returns all data that is loaded into a particular viewer. The only time this is not the case is when a subset is applied to multiple data. In that case, only the last data the subset applies to (going by layer order) will be returned with `data[‘Subset 1’]`.

https://github.com/spacetelescope/jdaviz/blob/46f0ae26605c2844716e472ae5234adfb1ea3ffa/jdaviz/app.py#L680

The next issue is that `app.get_subset_from_viewer()` uses the data that the subset is applied to (line 680 above) to calculate the bounds of that subset

https://github.com/spacetelescope/jdaviz/blob/46f0ae26605c2844716e472ae5234adfb1ea3ffa/jdaviz/app.py#L775-L776

https://github.com/spacetelescope/jdaviz/blob/46f0ae26605c2844716e472ae5234adfb1ea3ffa/jdaviz/app.py#L822

https://github.com/spacetelescope/jdaviz/blob/46f0ae26605c2844716e472ae5234adfb1ea3ffa/jdaviz/app.py#L826-L828

This means that if Subset 1 actually applies to multiple data, only the bounds that apply to the latest data will be used. I attempted to fix that last point in this PR

https://github.com/spacetelescope/jdaviz/pull/1886/files

But then we realized that these lines

https://github.com/spacetelescope/jdaviz/blob/3f851ec3525e59e31a90e7f7275f97352a93f19a/jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py#L147-L148

Append subregions to a subset, even if the user is just repositioning the subset. In order to fix this, we would have to include all the glue subset operators

https://jdaviz.readthedocs.io/en/latest/imviz/displayimages.html#defining-spatial-regions

shown at the bottom of those docs so we could arrange the subregions correctly.

We have discussed the possible solution of using a `SubsetCollection` type object to store all the data a subset is applied to, which would help with this particular problem. We have reached out to @astrofrog  if there is something like that that already exists but it may be something we have to create in jdaviz.

Example of a Subset applied to multiple spectra:
<img width="1412" alt="screen_shot_2022-11-22_at_3 23 38_pm" src="https://user-images.githubusercontent.com/7179333/207629336-c9620b58-2ca0-40c4-9019-b46a18831543.png">

**Comment by pllim** [comment]
I might be in the wrong rabbit hole but too some degree, I was able to traverse the tree after I added the following method in `CompositeSubsetState` over at glue core. Would save us from reverse-engineering the `__str__` outputs but as it is posted, the output requires some brain cells to parse.

https://github.com/glue-viz/glue/blob/583699bb13514473371e5059a6386e337677a92f/glue/core/subset.py#L1065

```diff
--- a/glue/core/subset.py
+++ b/glue/core/subset.py
@@ -1096,6 +1096,11 @@ class CompositeSubsetState(SubsetState):
         sym = OPSYM.get(self.op, self.op)
         return "(%s %s %s)" % (self.state1, sym, self.state2)

+    def traverse(self):
+        return (self.op,
+                self.state1 if not hasattr(self.state1, 'traverse') else self.state1.traverse(),
+                self.state2 if not hasattr(self.state2, 'traverse') else self.state2.traverse())
+
```

With the patch above at glue-core, you would get this in Specviz. Here, the composite subset is the second one I created and I used "AND" the first pass with 2 discontinuous regions, and then after that I got back and add XOR to the first sub-region. I don't know why each operator traversal returns in the format of `(op, some_state, (op, left, right))` as I was expecting something slightly more straightforward.

```python
>>> specviz.app.data_collection.subset_groups[1].subset_state.traverse()
(<function _operator.xor(a, b, /)>,
 <glue.core.subset.RangeSubsetState at 0x7f3ba140e440>,
 (<function _operator.xor(a, b, /)>,
  <glue.core.subset.RangeSubsetState at 0x7f3ba140ed10>,
  (<function _operator.or_(a, b, /)>,
   <glue.core.subset.RangeSubsetState at 0x7f3ba140e5f0>,
   (<function _operator.or_(a, b, /)>,
    <glue.core.subset.RangeSubsetState at 0x7f3ba14d58a0>,
    <glue.core.subset.RangeSubsetState at 0x7f3ba14d4d30>))))
```

Even without this patch, you can kinda see the traversal pattern by manually accessing the following:

```python
>>> specviz.app.data_collection.subset_groups[1].subset_state.op
>>> specviz.app.data_collection.subset_groups[1].subset_state.state1
>>> specviz.app.data_collection.subset_groups[1].subset_state.state2
>>> specviz.app.data_collection.subset_groups[1].subset_state.state2.op  # and so on
```

**Comment by kecnry** [comment]
Also affects the subset component's `selected_subset_mask` property (which probably should be refactored into a method which also takes the `spectrum1d` as input, similar to `selected_min_max` and is part of the issues causing #1910)

https://github.com/spacetelescope/jdaviz/blob/ecc2658d9b403dafc260826a29778014431c8e94/jdaviz/core/template_mixin.py#L958-L960

**Comment by astrofrog** [comment]
> as I was expecting something slightly more straightforward

Just to check, what would you have liked to see? Currently you are seeing the result of the very generic way of combining subsets which might not always be e.g. spatial subsets but you could have subsets defined also by e.g. a contour/threshold or even more wild subsets, an each subset in the expression might depend on different data attributes. What I mean is that not all subsets will always be geometrical subsets in the plane of the image being viewed, so not always possible to write in a simpler way. Having said that maybe we could investigate having a way for viewer classes to optionally be able to simplify subset definitions on the fly when they can be simplified.

**Comment by pllim** [comment]
@astrofrog , did the tag-up just now answer your question above?

**Comment by astrofrog** [comment]
Yep!

**Comment by javerbukh** [comment]
Hi @astrofrog here is an example of what we were talking about with the `traverse` method: https://github.com/glue-viz/glue/pull/2354

**Comment by javerbukh** [comment]
Here are the notes from the meeting that was had for this topic on 01/18/2023:

### How do we want subset information returned?
- as before (applied to each data and returned as a masked spectrum1d)
    - apply subsets keyword in get_data takes list of strings that are subset names and apply those subsets to the data
- as an astropy region showing all bounds of subset, even if compound?
- as a glue subset object (not sure how beneficial this would be)
- something else? 
    - define mask component for spectrum
    - do not want spectrum attaached to one particular mask
    - return as just a mask so that a user can use them as they wish in the notebook +
    - remove subsets from this method completely +
    - remove `get_viewer` from get subsets method (cubeviz case is special since a subset is spatial and also applied in spectrum-viewer as a collapsed function)
    - implement get_data_... at base helper class level so it can be public api
    - force get_data to require a data_label so data is returned along with associated subsets. This avoids returning as a dictionary where keys can be overwritten
    - get_data needs to know if data needs to be collapsed or not (use cls parameter, or overload cubeviz helper to have extra keyword arguement)
- method for accessing data labels present/visible in jdaviz viewers


### Path forward
We ended up deciding to do "something else", specifically implement a `get_data` method in the base helper class that returns data as type `cls` with subset `subset_to_apply` applied. This work is happening in #1984 . Implementing `get_subsets` will happen in a separate PR. Replacing `get_data_from_viewer` calls with `get_data` is also a separate effort.
